Synchronous walker moving

ye-luo opened the pull request:
Wait for the merge of #150 before reviewing this PR.

This PR introduce a synchronous walking driver.
multi_XXX routines handles multiple walkers simultaneously and by default fallback to the single walker implementation.

ye-luo committed:
Add experimental miniqmc_sync_move driver.

ye-luo committed:
Merge branch 'add-walker-loop' into walker-sync-move

ye-luo committed:
Enable synchronous walker moves.

ye-luo committed:
Prototype multi walker APIs.

ye-luo committed:
Construct a valid_mover_list.

ye-luo committed:
Enable multi_evaluateLog/acceptrestoreMove.

ye-luo committed:
Use references.

ye-luo committed:
Enable multi_evalGrad/ratioGrad.

ye-luo committed:
Merge branch 'add-walker-loop' into walker-sync-move

ye-luo committed:
Enable multi_evaluateGL.

prckent reviewed (changes requested):
(Placeholder change request to block merging pre #150 merge and before I/others add longer comments)

ye-luo committed:
Merge branch 'add-walker-loop' into walker-sync-move

ye-luo committed:
Merge branch 'develop' into walker-sync-move

ye-luo committed:
Merge branch 'develop' into walker-sync-move

ye-luo committed:
Use SPOSet in miniqmc_sync_move.

ye-luo committed:
Propagate timer change.

ye-luo committed:
Replace bool with int in vector for thread safety.

std::vector<bool> is special. The standard says
https://en.cppreference.com/w/cpp/container/vector_bool
Does not guarantee that different elements in the same container can be modified concurrently by different threads.
Namely writing to std::vector<bool> is not thread safe.

ye-luo committed:
Minor change.

ye-luo committed:
Merge branch 'develop' into walker-sync-move

ye-luo committed:
consistent rename WaveFunctionComponent.

ye-luo committed:
Add multi_evaluate_XXX for SPOs.

ye-luo committed:
Merge branch 'develop' into walker-sync-move

ye-luo committed:
Use multi_evaluate for SPO.

prckent commented:
Just a general comment to go on record with some of our discussions, not a review:

While we are experimenting in the miniapp I am fine for there to be multiple drivers. However, as we have discussed, it is a design goal that we can get to one architecture independent driver per QMC method in QMCPACK (and preferably "one" of as many things as possible below). Otherwise we will have a repeat of today's forkage, high costs to add architectures, unmaintainability, and challenge of encouraging contributions. I think everyone agrees that this is undesirable. The impact on human performance due to this design issue is much more important than (say) +-20% computational performance on any one platform. This is a design research topic, and one more important than which implementation strategies are eventually used (OMP, Kokkos, X, Y, Z).

prckent commented:
Approving since no other objections/comments received.
